File: ClientEngine/game/engine/Stage/UnitPart/UnitEvent/UnitEventControlTapScene.cpp
#include "UnitEventControlTapScene.h"
#include "../../../Controls/ControlEvent.h"
#include "../UnitBehavior.h"
#include "../../UnitInstance/UnitInstanceStandard.h"

UnitEventControlTapScene::UnitEventControlTapScene(SharedParams_t params) : UnitEventBase(params),
	TapSceneName(), DataName()
{
	IgnoreNode = false;
	IdentNode = true;
	FilterId = -1;
	TapSceneState = BUTTON_STATE_NONE;
}

UnitEventControlTapScene::~UnitEventControlTapScene(void)
{
}

// Выполняется ли условие
bool UnitEventControlTapScene::IsApprove( core::array<Event_t*>& events )
{
	for (u32 i = 0; i < events.size(); i++)
	{
		Event_t* eventInst = events[i];
		if(eventInst->EventId != EVENT_CONTROL_TAPSCENE) 
			continue;
		ControlTapSceneDesc_t* desc = (ControlTapSceneDesc_t*)eventInst->Params.CommonParam;

		if(desc->Name == TapSceneName && desc->State == TapSceneState)
		{
			if (IgnoreNode) return true;			
			if(FilterId != -1)
			{
				if(desc->Node == NULL) return false;
				bool validNode = ((desc->Node->getID() & FilterId) != 0);
				if(!validNode) return false;
			}
			ISceneNode* node = Behavior->GetSceneNode();
			if ((IdentNode && node == desc->Node) || (!IdentNode && node != desc->Node))
			{
				if(DataName != "")
				{
					UnitInstanceStandard* uInst = Behavior->GetUnitInstance();
					_DEBUG_BREAK_IF(uInst == NULL)
					uInst->SetData(DataName, desc->Node);					
				}
				return true;
			}
			return false;
		}
	}
	return false;
}

File: UnitEngine/Events/Control/UnitEventControlTapScene.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace UnitEngine.Events
{
    /// <summary>
    /// Описание события от клика по сцене
    /// </summary>
    [Serializable]
    public class UnitEventControlTapScene : UnitEventBase
    {
        /// <summary>
        /// Наименование контрола
        /// </summary>
        public string TapSceneName = string.Empty;

        /// <summary>
        /// Игнорировать модель
        /// </summary>
        public bool IgnoreNode = false;

        /// <summary>
        /// Модели должны совпадать
        /// </summary>
        public bool IdentNode = true;

        /// <summary>
        /// Индентификатор типа модели
        /// </summary>
        public int FilterId = -1;

        /// <summary>
        /// Сохранить модель в данных
        /// </summary>
        public string DataName = string.Empty;
        
        /// <summary>
        /// Состояние
        /// </summary>
        public UnitEventControlButtonState State = UnitEventControlButtonState.Down;

        public override string ToString()
        {
            return string.Format("TapScene [{0}] [{1}]", TapSceneName, State.ToString());
        }
    }
}
